client/tokenized-express-checkout/index.js

```diff
@@ -326,6 +326,21 @@ jQuery( ( $ ) => {
 				'wcpay.express-checkout.update-button-data',
 				'automattic/wcpay/express-checkout',
 				async () => {
+					// if the product cannot be added to cart (because of missing variation selection, etc),
+					// don't try to add it to the cart to get new data - the call will likely fail.
+					if (
+						getExpressCheckoutData( 'button_context' ) === 'product'
+					) {
+						const addToCartButton = $(
+							'.single_add_to_cart_button'
+						);
+
+						// First check if product can be added to cart.
+						if ( addToCartButton.is( '.disabled' ) ) {
+							return;
+						}
+					}
```

Comment:
This is the main change - it prevents from fetching new cart data if we're on a product page and the product can't be added to the cart.